### PR TITLE
fix: v0.8.1 — ANN rerank tuning + audit-driven bug fixes + test coverage

### PR DESCRIPTION
**Pull request by jyunming**
## Summary

Consolidated v0.8.1 release. Combines two streams of work into a single patch release:

1. **Performance — dimension-aware `rerank_factor`**: at d=1536 (OpenAI `text-embedding-3` dim), ANN search dropped from ~90 ms p50 to ~13 ms with *better* recall. Default oversampling factor steps down with vector dimension instead of using the previous fixed 20.
2. **Audit-driven bug fixes**: a 3-agent code audit surfaced ~80 candidate findings; the **9 verified critical/high items are fixed here**, and **5 previously untested modules get unit-test suites**.

No API changes. **Test count: 387 → 425** (+38; 2 long-running boundary tests marked `#[ignore]`).

This PR supersedes #86 — the dimension-aware rerank_factor commits from #86 are included here, and Copilot's actionable review feedback from #86 is folded in. PR #86 will be closed.

## Performance result (private bench, n=100k, ANN, rerank=True)

| Dataset (dim) | b=2 p50 before → after | b=2 R@1 before → after | b=4 p50 before → after | b=4 R@1 before → after |
|---|---|---|---|---|
| GloVe-200 | 19.5 → **2.4 ms** | 0.211 → **0.403** | 21.6 → **1.3 ms** | 0.366 → **0.422** |
| DBpedia-1536 | 90.0 → **13.3 ms** | 0.738 → **0.836** | 97.9 → **8.5 ms** | 0.822 → **0.850** |
| DBpedia-3072 | 43.2 → **23.8 ms** | 0.766 → **0.849** | 51.0 → **15.4 ms** | 0.816 → **0.861** |

Speed improved 1.8–17× and recall improved 2–19% across all dims.

## Critical fixes (from the audit)

- **WAL truncation logging** — partial writes now logged to stderr instead of silently dropped.
- **WAL oversized-payload guard** — `MAX_REASONABLE_PAYLOAD = 10 MiB` cap prevents gigabyte-allocs on corrupted length fields.
- **Graph NaN-score sanitization** — non-finite scorer outputs coerced to `NEG_INFINITY`; HNSW ranking is now deterministic.
- **Filter recursion-depth DoS guard** — `MAX_FILTER_DEPTH = 32` enforced; adversarial nested `$and`/`$or` no longer risks stack overflow.
- **IVF zero-cluster validation** — `create_coarse_index(n_clusters=0)` errors early; `probe(_, nprobe=0)` clamps to 1.
- **Quantizer dimension assertions** — explicit panic on mismatched `idx.len()` instead of UB via `get_unchecked`.

## High fixes

- **`internal_k` overflow guard** — `top_k * factor` now uses `saturating_mul` and caps at active count.
- **Bench version detection** — reads `pyproject.toml` directly (the documented source-of-truth) instead of the stale `importlib.metadata` cache that `maturin develop --release` doesn't refresh.
- **Python `Database.search` docstring** — updated to describe the new dimension-aware defaults.
- **Boundary recall continuity tests** — E2E tests at d=384/385 and d=1024/1025 verify recall doesn't cliff across the dimension cutoffs.

## Coverage suites

- **ivf.rs** — 9 tests (was 0)
- **wal.rs** — 5 replay edge cases
- **filter.rs** — 8 operator/depth coverage tests
- **prod.rs / mse.rs** — 4 numeric edge tests
- **graph.rs** — 3 beam-search edge tests

## Test plan

- [x] `cargo test --lib` — 423 pass + 2 ignored boundary tests (was 387)
- [x] Pre-commit gates green (fmt, check, lib tests, perf check)
- [x] Boundary tests pass: `cargo test -- --ignored boundary`
- [x] Private bench: no regression at d=1536
- [x] Public bench (`paper_recall_bench.py --update-readme --track`): regenerated for all 3 paper datasets

### CODE REVIEW

File: src/storage/engine/filter.rs [hunk 1]
```diff
@@ -690,3 +729,173 @@ mod or_fast_path_test {
...
+        let huge = json!(u64::MAX); // not representable in f64 → as_f64 = None → NaN
+        let meta = meta_one("count", huge);
+        let mut filter: HashMap<String, Value> = HashMap::new();
+        filter.insert("count".to_string(), json!({"$gt": 100}));
+        // u64::MAX as_f64() may return Some(huge_approx) on some serde_json
+        // versions; either way, the comparison must not panic.
+        let r = metadata_matches_filter(&meta, &filter);
+        // Expected behavior: either matches (because u64::MAX → f64 → > 100)
+        // or doesn't (NaN). Both are acceptable; pinning the no-panic contract.
+        let _ = r; // value not asserted — just must not crash.
```
[the remark this answers is not in the document]
Comment (jyunming) [reply src/storage/engine/filter.rs]: Addressed in commit c3cda24: renamed test to `big_int_metadata_with_numeric_op_compares_lossily`, added 3 actual assertions (u64::MAX > 100 is true, < 100 is false, adjacent huge ints don't panic), and updated the inline comment in `apply_comparison_op` to describe the lossy-conversion contract.

File: src/storage/graph.rs [hunk 2]
```diff
@@ -1081,4 +1092,100 @@ mod tests {
...
+        // No NaN scores in the output.
+        for (_id, score) in &r1 {
+            assert!(
+                score.is_finite(),
+                "result score {score} should be finite after NaN sanitization"
+            );
+        }
+    }
+
+    /// A2 corollary: scorer returning +Infinity for one node ranks it first.
```
[the remark this answers is not in the document]
Comment (jyunming) [reply src/storage/graph.rs]: Addressed in commit c3cda24: updated the doc comment to `+Inf is non-finite → sanitized to NEG_INFINITY → that node ranks LAST, not first` so it matches the assertion.

File: pyproject.toml [hunk 3]
```diff
@@ -4,7 +4,7 @@ build-backend = "maturin"
 
 [project]
 name = "tqdb"
-version = "0.8.0"
+version = "0.8.2"
```
[the remark this answers is not in the document]
Comment (jyunming) [reply pyproject.toml]: Not a repo convention. CLAUDE.md states the opposite: "`pyproject.toml` is the single source of truth for the version. To bump the version, update only `pyproject.toml`. The pre-commit hook detects any mismatch and updates `Cargo.toml` automatically." Bumping on the feature branch IS the documented workflow. Marking resolved.

File: Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "tqdb"
-version = "0.8.0"
+version = "0.8.2"
```
[the remark this answers is not in the document]
Comment (jyunming) [reply Cargo.toml]: Not a repo convention. CLAUDE.md says `Cargo.toml` is auto-synced by the pre-commit hook from `pyproject.toml`; never edit it manually. The bump here is the result of that sync, not a separate action. Marking resolved.
